File: public/code-examples/typescript/02-auto-compound/src/AutoCompounder.ts
/**
 * AutoCompounder - Core auto-compound implementation
 */

import { Connection, PublicKey, Keypair, Transaction } from '@solana/web3.js';
import { 
  claimRewardsSaros,
  addLiquiditySaros,
  stakeSaros,
  getPoolInfo,
  getUserPosition
} from './saros-sdk-mock';
import * as cron from 'node-cron';
import BN from 'bn.js';
import { logger } from './utils/logger';
import { GasOptimizer } from './GasOptimizer';
import { NotificationService } from './utils/notifications';

export interface AutoCompoundConfig {
  rpcUrl: string;
  privateKey: string;
  network?: 'devnet' | 'mainnet-beta';
  maxGasPrice?: number;
  emergencyStop?: () => Promise<boolean>;
}

export interface CompoundStrategy {
  poolAddress: PublicKey;
  strategy: 'LP' | 'STAKING' | 'FARMING';
  interval: number;
  minRewardThreshold: number;
  reinvestPercentage: number;
  maxSlippage?: number;
  emergencyWithdraw?: boolean;
}

export interface CompoundResult {
  success: boolean;
  rewardsHarvested: number;
  amountReinvested: number;
  newPositionValue: number;
  gasUsed: number;
  signature: string;
  timestamp: Date;
  error?: string;
}

export interface StartResult {
  success: boolean;
  poolAddress: string;
  strategy: string;
  interval: number;
  minThreshold: number;
  nextCompoundTime: string;
  error?: string;
}

export interface Statistics {
  totalCompounds: number;
  successfulCompounds: number;
  failedCompounds: number;
  successRate: number;
  totalRewardsHarvested: number;
  totalReinvested: number;
  totalGasSpent: number;
  netProfit: number;
  averageAPYBoost: number;
  lastCompoundTime?: Date;
}

interface ActivePool {
  poolAddress: PublicKey;
  strategy: CompoundStrategy;
  cronJob: cron.ScheduledTask;
  statistics: PoolStatistics;
}

interface PoolStatistics {
  compounds: number;
  totalHarvested: number;
  totalReinvested: number;
  totalGas: number;
  lastCompound?: Date;
  averageReward: number;
}

export class AutoCompounder {
  private connection: Connection;
  private wallet: Keypair;
  private activePools: Map<string, ActivePool> = new Map();
  private gasOptimizer: GasOptimizer;
  private notificationService: NotificationService;
  private globalStats: Statistics;
  private emergencyStopCheck?: () => Promise<boolean>;
  private maxGasPrice: number;

  constructor(config: AutoCompoundConfig) {
    this.connection = new Connection(config.rpcUrl, 'confirmed');
    // Handle different private key formats
    let secretKey: Uint8Array;
    if (config.privateKey.startsWith('[') && config.privateKey.endsWith(']')) {
      // JSON array format
      const keyArray = JSON.parse(config.privateKey);
      secretKey = Uint8Array.from(keyArray);
    } else {
      // Assume base64 format
      secretKey = new Uint8Array(Buffer.from(config.privateKey, 'base64'));
    }
    this.wallet = Keypair.fromSecretKey(secretKey);
    this.gasOptimizer = new GasOptimizer(this.connection);
    this.notificationService = new NotificationService();
    this.maxGasPrice = config.maxGasPrice || 0.01;
    this.emergencyStopCheck = config.emergencyStop;
    
    this.globalStats = {
      totalCompounds: 0,
      successfulCompounds: 0,
      failedCompounds: 0,
      successRate: 0,
      totalRewardsHarvested: 0,
      totalReinvested: 0,
      totalGasSpent: 0,
      netProfit: 0,
      averageAPYBoost: 0
    };

    logger.info('AutoCompounder initialized');
    logger.info(`Wallet: ${this.wallet.publicKey.toString()}`);
  }

  /**
   * Start auto-compounding for a pool
   */
  async start(strategy: CompoundStrategy): Promise<StartResult> {
    const poolKey = strategy.poolAddress.toString();
    
    try {
      // Check if already active
      if (this.activePools.has(poolKey)) {
        logger.warn(`Pool ${poolKey} already active`);
        return {
          success: false,
          poolAddress: poolKey,
          strategy: strategy.strategy,
          interval: strategy.interval,
          minThreshold: strategy.minRewardThreshold,
          nextCompoundTime: '',
          error: 'Pool already active'
        };
      }

      // Validate pool exists
      const poolInfo = await this.getPoolInfo(strategy.poolAddress);
      if (!poolInfo) {
        throw new Error('Pool not found');
      }

      // Create cron schedule
      const cronExpression = this.intervalToCron(strategy.interval);
      const cronJob = cron.schedule(cronExpression, async () => {
        await this.executeCompound(strategy);
      });

      // Store active pool
      const activePool: ActivePool = {
        poolAddress: strategy.poolAddress,
        strategy,
        cronJob,
        statistics: {
          compounds: 0,
          totalHarvested: 0,
          totalReinvested: 0,
          totalGas: 0,
          averageReward: 0
        }
      };

      this.activePools.set(poolKey, activePool);
      cronJob.start();

      // Calculate next compound time
      const nextCompound = new Date(Date.now() + strategy.interval);

      logger.info(`✅ Auto-compound started for pool ${poolKey}`);
      logger.info(`  Strategy: ${strategy.strategy}`);
      logger.info(`  Interval: ${strategy.interval}ms`);
      logger.info(`  Next compound: ${nextCompound.toLocaleString()}`);

      // Send notification
      await this.notificationService.send({
        type: 'COMPOUND_STARTED',
        pool: poolKey,
        strategy: strategy.strategy,
        interval: strategy.interval
      });

      return {
        success: true,
        poolAddress: poolKey,
        strategy: strategy.strategy,
        interval: strategy.interval,
        minThreshold: strategy.minRewardThreshold,
        nextCompoundTime: nextCompound.toLocaleString()
      };

    } catch (error: any) {
      logger.error(`Failed to start auto-compound: ${error.message}`);
      
      return {
        success: false,
        poolAddress: poolKey,
        strategy: strategy.strategy,
        interval: strategy.interval,
        minThreshold: strategy.minRewardThreshold,
        nextCompoundTime: '',
        error: error.message
      };
    }
  }

  /**
   * Stop auto-compounding for a pool
   */
  async stop(poolAddress: PublicKey): Promise<boolean> {
    const poolKey = poolAddress.toString();
    const activePool = this.activePools.get(poolKey);
    
    if (!activePool) {
      logger.warn(`Pool ${poolKey} not active`);
      return false;
    }

    // Stop cron job
    activePool.cronJob.stop();
    
    // Remove from active pools
    this.activePools.delete(poolKey);

    logger.info(`Auto-compound stopped for pool ${poolKey}`);
    
    // Send notification
    await this.notificationService.send({
      type: 'COMPOUND_STOPPED',
      pool: poolKey,
      statistics: activePool.statistics
    });

    return true;
  }

  /**
   * Manually trigger compound for a pool
   */
  async compoundNow(poolAddress: PublicKey): Promise<CompoundResult> {
    const poolKey = poolAddress.toString();
    const activePool = this.activePools.get(poolKey);
    
    if (!activePool) {
      // Create temporary strategy for manual compound
      const strategy: CompoundStrategy = {
        poolAddress,
        strategy: 'LP',
        interval: 0,
        minRewardThreshold: 0,
        reinvestPercentage: 100
      };
      
      return this.executeCompound(strategy);
    }

    return this.executeCompound(activePool.strategy);
  }

  /**
   * Execute compound operation
   */
  private async executeCompound(strategy: CompoundStrategy): Promise<CompoundResult> {
    const startTime = Date.now();
    const poolKey = strategy.poolAddress.toString();
    
    try {
      logger.info(`🔄 Executing compound for pool ${poolKey}`);

      // Check emergency stop
      if (this.emergencyStopCheck && await this.emergencyStopCheck()) {
        logger.warn('Emergency stop triggered, skipping compound');
        return this.createFailedResult('Emergency stop triggered');
      }

      // Check gas price
      const gasPrice = await this.getCurrentGasPrice();
      if (gasPrice > this.maxGasPrice) {
        logger.warn(`Gas price too high: ${gasPrice} > ${this.maxGasPrice}`);
        return this.createFailedResult('Gas price too high');
      }

      // Get current position and rewards
      const position = await this.getUserPosition(strategy.poolAddress);
      const pendingRewards = await this.getPendingRewards(strategy.poolAddress);

      logger.info(`  Position: ${position.amount}`);
      logger.info(`  Pending rewards: ${pendingRewards}`);

      // Check minimum threshold
      if (pendingRewards < strategy.minRewardThreshold) {
        logger.info(`Rewards below threshold: ${pendingRewards} < ${strategy.minRewardThreshold}`);
        return this.createFailedResult('Rewards below threshold');
      }

      // Step 1: Harvest rewards
      const harvestResult = await this.harvestRewards(strategy.poolAddress);
      if (!harvestResult.success) {
        throw new Error('Failed to harvest rewards');
      }

      // Step 2: Calculate reinvest amount
      const reinvestAmount = (pendingRewards * strategy.reinvestPercentage) / 100;
      const keepAmount = pendingRewards - reinvestAmount;

      logger.info(`  Reinvesting: ${reinvestAmount} (${strategy.reinvestPercentage}%)`);
      if (keepAmount > 0) {
        logger.info(`  Keeping: ${keepAmount}`);
      }

      // Step 3: Reinvest based on strategy
      let reinvestResult;
      switch (strategy.strategy) {
        case 'LP':
          reinvestResult = await this.reinvestLP(
            strategy.poolAddress,
            reinvestAmount,
            strategy.maxSlippage || 1.0
          );
          break;
          
        case 'STAKING':
          reinvestResult = await this.reinvestStaking(
            strategy.poolAddress,
            reinvestAmount
          );
          break;
          
        case 'FARMING':
          reinvestResult = await this.reinvestFarming(
            strategy.poolAddress,
            reinvestAmount,
            strategy.maxSlippage || 1.0
          );
          break;
          
        default:
          throw new Error(`Unknown strategy: ${strategy.strategy}`);
      }

      if (!reinvestResult.success) {
        throw new Error('Failed to reinvest');
      }

      // Step 4: Get new position value
      const newPosition = await this.getUserPosition(strategy.poolAddress);
      const gasUsed = (Date.now() - startTime) * 0.000001; // Simplified gas calculation

      // Update statistics
      this.updateStatistics(poolKey, {
        success: true,
        rewardsHarvested: pendingRewards,
        amountReinvested: reinvestAmount,
        gasUsed
      });

      // Send notification
      await this.notificationService.send({
        type: 'COMPOUND_SUCCESS',
        pool: poolKey,
        rewards: pendingRewards,
        reinvested: reinvestAmount,
        newPosition: newPosition.amount
      });

      const result: CompoundResult = {
        success: true,
        rewardsHarvested: pendingRewards,
        amountReinvested: reinvestAmount,
        newPositionValue: newPosition.amount,
        gasUsed,
        signature: reinvestResult.signature,
        timestamp: new Date()
      };

      logger.info(`✅ Compound successful for pool ${poolKey}`);
      logger.info(`  Harvested: ${pendingRewards}`);
      logger.info(`  Reinvested: ${reinvestAmount}`);
      logger.info(`  Gas used: ${gasUsed} SOL`);

      return result;

    } catch (error: any) {
      logger.error(`Compound failed for pool ${poolKey}: ${error.message}`);
      
      // Update statistics
      this.updateStatistics(poolKey, {
        success: false,
        rewardsHarvested: 0,
        amountReinvested: 0,
        gasUsed: 0
      });

      // Send notification
      await this.notificationService.send({
        type: 'COMPOUND_FAILED',
        pool: poolKey,
        error: error.message
      });

      return this.createFailedResult(error.message);
    }
  }

  /**
   * Harvest rewards from pool
   */
  private async harvestRewards(poolAddress: PublicKey): Promise<any> {
    // Implement actual harvest logic using Saros SDK
    // This is a placeholder
    return {
      success: true,
      signature: 'harvest_tx_signature'
    };
  }

  /**
   * Reinvest into LP position
   */
  private async reinvestLP(
    poolAddress: PublicKey,
    amount: number,
    maxSlippage: number
  ): Promise<any> {
    // Implement LP reinvestment using Saros SDK
    // This is a placeholder
    return {
      success: true,
      signature: 'reinvest_lp_tx_signature'
    };
  }

  /**
   * Reinvest into staking position
   */
  private async reinvestStaking(
    poolAddress: PublicKey,
    amount: number
  ): Promise<any> {
    // Implement staking reinvestment using Saros SDK
    // This is a placeholder
    return {
      success: true,
      signature: 'reinvest_stake_tx_signature'
    };
  }

  /**
   * Reinvest into farming position
   */
  private async reinvestFarming(
    poolAddress: PublicKey,
    amount: number,
    maxSlippage: number
  ): Promise<any> {
    // Implement farming reinvestment using Saros SDK
    // This is a placeholder
    return {
      success: true,
      signature: 'reinvest_farm_tx_signature'
    };
  }

  /**
   * Get pool information
   */
  private async getPoolInfo(poolAddress: PublicKey): Promise<any> {
    // Implement using Saros SDK
    // This is a placeholder
    return {
      exists: true,
      tvl: 1000000,
      apy: 50
    };
  }

  /**
   * Get user position
   */
  private async getUserPosition(poolAddress: PublicKey): Promise<any> {
    // Implement using Saros SDK
    // This is a placeholder
    return {
      amount: 1000,
      rewards: 10
    };
  }

  /**
   * Get pending rewards
   */
  private async getPendingRewards(poolAddress: PublicKey): Promise<number> {
    // Implement using Saros SDK
    // This is a placeholder
    return Math.random() * 10 + 1; // Random rewards for demo
  }

  /**
   * Get current gas price
   */
  async getCurrentGasPrice(): Promise<number> {
    return this.gasOptimizer.getCurrentGasPrice();
  }

  /**
   * Convert interval to cron expression
   */
  private intervalToCron(intervalMs: number): string {
    const minutes = Math.floor(intervalMs / 60000);
    
    if (minutes < 60) {
      return `*/${minutes} * * * *`; // Every X minutes
    }
    
    const hours = Math.floor(minutes / 60);
    if (hours < 24) {
      return `0 */${hours} * * *`; // Every X hours
    }
    
    const days = Math.floor(hours / 24);
    return `0 0 */${days} * *`; // Every X days
  }

  /**
   * Update statistics
   */
  private updateStatistics(
    poolKey: string,
    result: {
      success: boolean;
      rewardsHarvested: number;
      amountReinvested: number;
      gasUsed: number;
    }
  ): void {
    // Update global stats
    this.globalStats.totalCompounds++;
    
    if (result.success) {
      this.globalStats.successfulCompounds++;
      this.globalStats.totalRewardsHarvested += result.rewardsHarvested;
      this.globalStats.totalReinvested += result.amountReinvested;
      this.globalStats.totalGasSpent += result.gasUsed;
    } else {
      this.globalStats.failedCompounds++;
    }
    
    this.globalStats.successRate = 
      (this.globalStats.successfulCompounds / this.globalStats.totalCompounds) * 100;
    
    this.globalStats.netProfit = 
      this.globalStats.totalRewardsHarvested - this.globalStats.totalGasSpent;
    
    this.globalStats.lastCompoundTime = new Date();

    // Update pool-specific stats
    const activePool = this.activePools.get(poolKey);
    if (activePool) {
      activePool.statistics.compounds++;
      activePool.statistics.totalHarvested += result.rewardsHarvested;
      activePool.statistics.totalReinvested += result.amountReinvested;
      activePool.statistics.totalGas += result.gasUsed;
      activePool.statistics.lastCompound = new Date();
      
      activePool.statistics.averageReward = 
        activePool.statistics.totalHarvested / activePool.statistics.compounds;
    }
  }

  /**
   * Create failed result
   */
  private createFailedResult(error: string): CompoundResult {
    return {
      success: false,
      rewardsHarvested: 0,
      amountReinvested: 0,
      newPositionValue: 0,
      gasUsed: 0,
      signature: '',
      timestamp: new Date(),
      error
    };
  }

  /**
   * Get statistics
   */
  async getStatistics(): Promise<Statistics> {
    return { ...this.globalStats };
  }

  /**
   * Get active pools
   */
  async getActivePools(): Promise<PublicKey[]> {
    return Array.from(this.activePools.keys()).map(key => new PublicKey(key));
  }

  /**
   * Set emergency stop function
   */
  setEmergencyStop(checkFn: () => Promise<boolean>): void {
    this.emergencyStopCheck = checkFn;
    logger.info('Emergency stop function configured');
  }

  /**
   * Stop all active compounds
   */
  async stopAll(): Promise<void> {
    logger.info('Stopping all active compounds...');
    
    for (const [poolKey, activePool] of this.activePools.entries()) {
      activePool.cronJob.stop();
      logger.info(`Stopped compound for pool ${poolKey}`);
    }
    
    this.activePools.clear();
    logger.info('All compounds stopped');
  }
}

File: code-examples/rust/02-auto-compound/src/auto_compounder.rs
use anyhow::Result;
use chrono::Utc;
use dashmap::DashMap;
use log::{error, info, warn};
use solana_client::rpc_client::RpcClient;
use solana_sdk::{
    pubkey::Pubkey,
    signature::{Keypair, Signer},
};
use std::{str::FromStr, sync::Arc, time::Duration};
use tokio::sync::RwLock;
use tokio_cron_scheduler::{Job, JobScheduler};

use crate::{
    compound_strategy::CompoundStrategy,
    gas_optimizer::GasOptimizer,
    notification_service::NotificationService,
    position_monitor::PositionMonitor,
    statistics::StatisticsManager,
    types::*,
};

pub struct ActiveStrategy {
    pub config: CompoundStrategyConfig,
    pub strategy: Arc<CompoundStrategy>,
    pub job_id: Option<uuid::Uuid>,
    pub statistics: PoolStatistics,
}

/// Main auto-compounder that manages multiple compound strategies
pub struct AutoCompounder {
    rpc_client: Arc<RpcClient>,
    wallet: Arc<Keypair>,
    config: AutoCompoundConfig,
    active_strategies: Arc<DashMap<String, ActiveStrategy>>,
    scheduler: Arc<JobScheduler>,
    gas_optimizer: Arc<GasOptimizer>,
    notification_service: Arc<NotificationService>,
    position_monitor: Arc<PositionMonitor>,
    statistics_manager: Arc<RwLock<StatisticsManager>>,
    start_time: chrono::DateTime<Utc>,
}

impl AutoCompounder {
    /// Create a new AutoCompounder instance
    pub async fn new(config: AutoCompoundConfig) -> Result<Self> {
        // Initialize RPC client
        let rpc_client = Arc::new(RpcClient::new(&config.rpc_url));

        // Initialize wallet from private key
        let wallet = Arc::new(Self::parse_private_key(&config.private_key)?);

        // Initialize scheduler
        let scheduler = Arc::new(JobScheduler::new().await?);
        scheduler.start().await?;

        // Initialize components
        let gas_optimizer = Arc::new(GasOptimizer::new(rpc_client.clone()));
        let notification_service = Arc::new(NotificationService::new(&config));
        let position_monitor = Arc::new(PositionMonitor::new(rpc_client.clone()));
        let statistics_manager = Arc::new(RwLock::new(StatisticsManager::new()));

        info!("🔑 Wallet address: {}", wallet.pubkey());
        info!("🌐 Network: {}", config.network);
        info!("📡 RPC URL: {}", config.rpc_url);

        Ok(Self {
            rpc_client,
            wallet,
            config,
            active_strategies: Arc::new(DashMap::new()),
            scheduler,
            gas_optimizer,
            notification_service,
            position_monitor,
            statistics_manager,
            start_time: Utc::now(),
        })
    }

    /// Start a compound strategy
    pub async fn start_strategy(&mut self, config: CompoundStrategyConfig) -> Result<StartResult> {
        let pool_key = config.pool_address.to_string();

        // Check if strategy is already active
        if self.active_strategies.contains_key(&pool_key) {
            warn!("Strategy for pool {} is already active", pool_key);
            return Ok(StartResult {
                success: false,
                pool_address: pool_key,
                strategy_type: config.strategy_type.to_string(),
                interval_ms: config.interval_ms,
                min_threshold: config.min_reward_threshold,
                next_compound_time: "".to_string(),
                error: Some("Strategy already active".to_string()),
            });
        }

        // Validate pool exists and get initial position
        let position = match self.position_monitor.get_position(config.pool_address).await {
            Ok(pos) => pos,
            Err(e) => {
                error!("Failed to get position for pool {}: {}", pool_key, e);
                return Ok(StartResult {
                    success: false,
                    pool_address: pool_key,
                    strategy_type: config.strategy_type.to_string(),
                    interval_ms: config.interval_ms,
                    min_threshold: config.min_reward_threshold,
                    next_compound_time: "".to_string(),
                    error: Some(format!("Failed to validate pool: {}", e)),
                });
            }
        };

        info!("📊 Initial position for {}: {:.6} LP tokens", pool_key, position.lp_token_amount);

        // Create compound strategy
        let strategy = Arc::new(CompoundStrategy::new(
            config.clone(),
            self.rpc_client.clone(),
            self.wallet.clone(),
            self.gas_optimizer.clone(),
            self.notification_service.clone(),
        ));

        // Create cron schedule based on interval
        let cron_expression = Self::interval_to_cron(config.interval_ms);
        info!("⏰ Scheduling with cron expression: {}", cron_expression);

        // Clone necessary values for the job
        let strategy_clone = strategy.clone();
        let statistics_manager = self.statistics_manager.clone();
        let pool_address = config.pool_address;

        // Create scheduled job
        let job = Job::new_async(cron_expression.as_str(), move |_uuid, _l| {
            let strategy = strategy_clone.clone();
            let stats_manager = statistics_manager.clone();
            
            Box::pin(async move {
                info!("🔄 Executing scheduled compound for pool: {}", pool_address);
                
                match strategy.execute_compound().await {
                    Ok(result) => {
                        info!("✅ Compound successful for pool {}", pool_address);
                        
                        // Update statistics
                        let mut stats = stats_manager.write().await;
                        stats.record_compound_result(&pool_address.to_string(), &result).await;
                    }
                    Err(e) => {
                        error!("❌ Compound failed for pool {}: {}", pool_address, e);
                        
                        // Record failed compound
                        let mut stats = stats_manager.write().await;
                        let failed_result = CompoundResult {
                            success: false,
                            rewards_harvested: 0.0,
                            amount_reinvested: 0.0,
                            new_position_value: 0.0,
                            gas_used: 0.0,
                            transaction_signature: "".to_string(),
                            timestamp: Utc::now(),
                            error: Some(e.to_string()),
                        };
                        stats.record_compound_result(&pool_address.to_string(), &failed_result).await;
                    }
                }
            })
        })?;

        // Add job to scheduler
        let job_id = self.scheduler.add(job).await?;

        // Calculate next compound time
        let next_compound = Utc::now() + chrono::Duration::milliseconds(config.interval_ms as i64);

        // Store active strategy
        let active_strategy = ActiveStrategy {
            config: config.clone(),
            strategy,
            job_id: Some(job_id),
            statistics: PoolStatistics {
                pool_address: config.pool_address,
                ..Default::default()
            },
        };

        self.active_strategies.insert(pool_key.clone(), active_strategy);

        // Send notification
        self.notification_service.send_notification(NotificationEvent {
            event_type: NotificationEventType::CompoundStarted,
            pool_address: pool_key.clone(),
            message: format!("Auto-compound started for {} strategy", config.strategy_type),
            data: serde_json::json!({
                "strategy_type": config.strategy_type.to_string(),
                "interval_ms": config.interval_ms,
                "min_threshold": config.min_reward_threshold,
            }),
            timestamp: Utc::now(),
        }).await;

        info!("✅ Auto-compound started for pool {}", pool_key);
        info!("   Strategy: {}", config.strategy_type);
        info!("   Interval: {}ms", config.interval_ms);
        info!("   Next compound: {}", next_compound.format("%Y-%m-%d %H:%M:%S UTC"));

        Ok(StartResult {
            success: true,
            pool_address: pool_key,
            strategy_type: config.strategy_type.to_string(),
            interval_ms: config.interval_ms,
            min_threshold: config.min_reward_threshold,
            next_compound_time: next_compound.format("%Y-%m-%d %H:%M:%S UTC").to_string(),
            error: None,
        })
    }

    /// Stop a compound strategy
    pub async fn stop_strategy(&mut self, pool_address: Pubkey) -> Result<bool> {
        let pool_key = pool_address.to_string();

        match self.active_strategies.remove(&pool_key) {
            Some((_, active_strategy)) => {
                // Remove job from scheduler
                if let Some(job_id) = active_strategy.job_id {
                    self.scheduler.remove(&job_id).await?;
                }

                // Send notification
                self.notification_service.send_notification(NotificationEvent {
                    event_type: NotificationEventType::CompoundStopped,
                    pool_address: pool_key.clone(),
                    message: "Auto-compound stopped".to_string(),
                    data: serde_json::json!({
                        "statistics": active_strategy.statistics
                    }),
                    timestamp: Utc::now(),
                }).await;

                info!("🛑 Auto-compound stopped for pool {}", pool_key);
                Ok(true)
            }
            None => {
                warn!("No active strategy found for pool {}", pool_key);
                Ok(false)
            }
        }
    }

    /// Manually trigger compound for a specific pool
    pub async fn compound_now(&self, pool_address: Pubkey) -> Result<CompoundResult> {
        let pool_key = pool_address.to_string();

        match self.active_strategies.get(&pool_key) {
            Some(active_strategy) => {
                info!("🔄 Manual compound triggered for pool {}", pool_key);
                active_strategy.strategy.execute_compound().await
            }
            None => {
                // Create temporary strategy for one-time compound
                let temp_config = CompoundStrategyConfig {
                    pool_address,
                    strategy_type: StrategyType::LP,
                    interval_ms: 0,
                    min_reward_threshold: 0.0,
                    reinvest_percentage: 100,
                    max_slippage: Some(1.0),
                    emergency_withdraw: false,
                };

                let temp_strategy = CompoundStrategy::new(
                    temp_config,
                    self.rpc_client.clone(),
                    self.wallet.clone(),
                    self.gas_optimizer.clone(),
                    self.notification_service.clone(),
                );

                temp_strategy.execute_compound().await
            }
        }
    }

    /// Get global statistics
    pub async fn get_global_statistics(&self) -> Result<GlobalStatistics> {
        let stats_manager = self.statistics_manager.read().await;
        let mut stats = stats_manager.get_global_statistics().await;
        
        // Calculate uptime
        let uptime = Utc::now() - self.start_time;
        stats.uptime_hours = uptime.num_minutes() as f64 / 60.0;
        
        Ok(stats)
    }

    /// Get statistics for a specific pool
    pub async fn get_pool_statistics(&self, pool_address: Pubkey) -> Result<Option<PoolStatistics>> {
        let stats_manager = self.statistics_manager.read().await;
        Ok(stats_manager.get_pool_statistics(&pool_address.to_string()).await)
    }

    /// Get list of active pools
    pub async fn get_active_pools(&self) -> Vec<Pubkey> {
        self.active_strategies
            .iter()
            .map(|entry| entry.key().parse::<Pubkey>().unwrap())
            .collect()
    }

    /// Stop all active strategies
    pub async fn stop_all(&mut self) -> Result<()> {
        info!("🛑 Stopping all active strategies...");
        
        let pool_keys: Vec<String> = self.active_strategies
            .iter()
            .map(|entry| entry.key().clone())
            .collect();

        for pool_key in pool_keys {
            if let Ok(pool_address) = pool_key.parse::<Pubkey>() {
                if let Err(e) = self.stop_strategy(pool_address).await {
                    error!("Failed to stop strategy for pool {}: {}", pool_key, e);
                }
            }
        }

        // Stop scheduler - we'll just let it drop naturally since we can't get mutable access

        info!("✅ All strategies stopped");
        Ok(())
    }

    /// Parse private key from various formats
    fn parse_private_key(private_key: &str) -> Result<Keypair> {
        // Try base58 format first
        if let Ok(decoded) = bs58::decode(private_key).into_vec() {
            if decoded.len() == 64 {
                return Ok(Keypair::from_bytes(&decoded)?);
            }
        }

        // Try JSON array format
        if private_key.starts_with('[') && private_key.ends_with(']') {
            let key_array: Vec<u8> = serde_json::from_str(private_key)?;
            if key_array.len() == 64 {
                return Ok(Keypair::from_bytes(&key_array)?);
            }
        }

        Err(anyhow::anyhow!("Invalid private key format"))
    }

    /// Convert interval in milliseconds to cron expression
    fn interval_to_cron(interval_ms: u64) -> String {
        let minutes = interval_ms / 60000;

        if minutes < 60 {
            format!("0/{} * * * *", minutes) // Every X minutes
        } else {
            let hours = minutes / 60;
            if hours < 24 {
                format!("0 0/{} * * *", hours) // Every X hours
            } else {
                let days = hours / 24;
                format!("0 0 0/{} * *", days) // Every X days
            }
        }
    }
}